### PR TITLE
Fix the flickering and other lighting issues related to zone entities (part 1)

### PR DESCRIPTION
**Pull request by samcake**
## Feedback welcome

This PR try to address specifically these bugs:
https://highfidelity.fogbugz.com/f/cases/3447/content-is-flashing
https://highfidelity.fogbugz.com/f/cases/2328/Zone-flickering

Instead of relying on a "zone Stack State" to assign the different components of rendering defined by  zone entities, we changed the architecture so that each zone has a corresponding render item. 
These items are collected and managed in one rendering job (once per frame) to produce the expected result, which is to pick the key light, ambient light and skybox applied at the avatar position depending on the zone stack.

In order to understand the Zone stack principle we also developed a debugging tool to be able to visualize the component values (key light, ambient map, skybox) and validate the behavior.

This the first part of that work, which produces the expected result as in master (but without the flickering in lighting or skyboxes when moving around in between zones) .
I have disabled the "Stage" feature of the zone entity which i m predicting is never used.
We will fix it in a second pr.

The procedural skyboxes are not displayed in the debugging tool, this would be good to fix it in a follow up pr.

## TEST PLAN
You could go in the blue domain which has an interesting zone configuration making it easy to debug.
Run this PR and run  the modified version of the debugDeferredLighting.js script.
There is a new check box "zone stack" which enable a new debug job that displays in front of you the 3 stacks of components modified by the zone entities in the domain:
Key Light, Ambient, Skybox
As you navigate the world and various zones, the stack should reflect the current values picked for each component (always the front one) as well as the other component values coming from the other zones in the stack.
When NOT in  a zone, the default values for the components are displayed

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/139/HighFidelity-Beta-PR10537-c256893575e9935ced4bb789318b5c6aea13247c.exe

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -337,7 +337,7 @@ class DeadlockWatchdogThread : public QThread {
                 // Don't actually crash in debug builds, in case this apparent deadlock is simply from
                 // the developer actively debugging code
                 #ifdef NDEBUG
-                    deadlockDetectionCrash();
+               //     deadlockDetectionCrash();
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: cleanup

File: libraries/entities-renderer/src/RenderableZoneEntityItem.cpp [hunk 2]
```diff
@@ -67,8 +143,18 @@ bool RenderableZoneEntityItem::setProperties(const EntityItemProperties& propert
 
 void RenderableZoneEntityItem::somethingChangedNotification() {
     DependencyManager::get<EntityTreeRenderer>()->updateZone(_id);
+
+    // If graphics elements are changed, we need to update the render items
+    if (_keyLightPropertiesChanged || _backgroundPropertiesChanged || _stagePropertiesChanged || _skyboxPropertiesChanged) {
+
+        notifyChangedRenderItem();
```
Comment (ZappoMan) [inline libraries/entities-renderer/src/RenderableZoneEntityItem.cpp]: what about if the bounds or position changes? don't we also need to adjust the render item?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/156/HighFidelity-Beta-PR10537-044110e4f3f324b2dc02de99528ad7d167024ef2.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/158/HighFidelity-Beta-PR10537-575d4abdf710a7381cfe846d1dee7fed3252449f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/158/HighFidelity-Beta-PR10537-575d4abdf710a7381cfe846d1dee7fed3252449f.dmg

File: libraries/entities-renderer/src/RenderableZoneEntityItem.cpp (hunk 2 above)
Comment (samcake) [inline libraries/entities-renderer/src/RenderableZoneEntityItem.cpp]: THis is correct, good catch

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/169/HighFidelity-Beta-PR10537-3eb95d2d193da2436a458e6f1b6d3a77f5c92268.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/169/HighFidelity-Beta-PR10537-3eb95d2d193da2436a458e6f1b6d3a77f5c92268.dmg

**Comment by Nex-Pro** [comment]
Testing this Win10 and Vive

**Comment by Nex-Pro** [comment]
@samcake blue took long time to begin to load ( was showing no servers ) Just after it started to load Interface crashed. I had seen the build platform ( white grid )  splat posted.

**Comment by Nex-Pro** [comment]
@samcake  **This looks good** Let me know if any other tests are needed. Merci beaucoup.

I created some nested zones. smallest of which intersected a zone smaller than the main zone in dev-playa.

As I moved thru the smaller zone which had no skybox set it simply inherited the skybox of the next nearest zone.

No flickering of the sky was seen.


Zone stack seems to work ok but a little strange looking. ( green keylight used ) 
hifi-snap-by-piper-peppercorn-on-2017-05-25_13-34-56

zone_stack

**Observations/feedback**

Also made two identically sized and positioned zones each with different keylight colour.  first was white keylight second was green.

Every time the two co-incident zones were entered the older ( White ) zone had priority and the other was never seen.

I then moved the green zone one meter to the side.  Entering it  correctly showed the green keylight and moving to where the white zone started immediately activated the white zone. Again the older zone had priority.

**Comment by samcake** [comment]
Great Review @Nex-Pro, thanks !

About your observation, i agree, the rules for switching between zones are based on the size of the zone.
When moving at the border between 2 zones of the same size, the rule of priority are based on the history order (basically the order in which they have been created but in your local client!)
THis is non determistic and could be a separate bug that we could address.

About your line "Zone stack seems to work ok but a little strange looking. ( green keylight used )"
I don't really understand your point here ?

thanks

**Comment by samcake** [comment]
Added the fix for the case when bound is changing and not the rest of the properties

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/182/HighFidelity-Beta-PR10537-cba913c06828dea77cc153b6577e8f0e8ea1502a.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/182/HighFidelity-Beta-PR10537-cba913c06828dea77cc153b6577e8f0e8ea1502a.dmg

**Comment by Nex-Pro** [comment]
Will re-test this

**Comment by Nex-Pro** [comment]
@samcake 

Interesting issue. placed the mars rocks item from market. Then got  the full moon slybox which always appears at 0,0,0  moved it to my location  then resized i to 30 30 30
it did not show up until I walked out of it and back.

I arranged one large zone  ( mars rocks ) then one of 30 x 30 x 30  ( full moon ) and one of 10 10 10 (just skybox colour ) the smaller intersected the 30 30 30   the lighting of the smaller one would only show up in the overlap between it and the 30 30 30
but the skybox colour ( pink sky)  works in the whole 10 10 10 .   I don't think this is intended.

**Comment by samcake** [comment]
this sounds like a bug, can i find you somewhere to debug ?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/222/HighFidelity-Beta-PR10537-882df7391b9b48e4ecdb39b3e4427d9d03bf0b4b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/222/HighFidelity-Beta-PR10537-882df7391b9b48e4ecdb39b3e4427d9d03bf0b4b.dmg

**Comment by samcake** [comment]
II m pretty confident that we are QA pass
since i plan to have a part2 anyway to adress the hidden sunskymodel feature i can have a follow up on this
i m going to merge
THanks Piper

**Pull request merged**